cpp/src/decisiontree/batched-levelalgo/builder_base.cuh

```diff
@@ -223,243 +289,174 @@ struct Builder {
+    ML::PUSH_RANGE("Builder::computeSplit @builder_base.cuh [batched-levelalgo]");
+    auto nbins    = params.n_bins;
+    auto nclasses = input.numOutputs;
+    auto colBlks  = std::min(n_blks_for_cols, input.nSampledCols - col);
+
+    auto smemSize = computeSplitSmemSize();
+    dim3 grid(total_blocks, colBlks, 1);
+    int nHistBins = large_blocks * nbins * colBlks * nclasses;
+    CUDA_CHECK(cudaMemsetAsync(hist, 0, sizeof(BinT) * nHistBins, handle.get_stream()));
+    ML::PUSH_RANGE("computeSplitClassificationKernel @builder_base.cuh [batched-levelalgo]");
     ObjectiveT objective(input.numOutputs, params.min_impurity_decrease, params.min_samples_leaf);
     computeSplitKernel<DataT, LabelT, IdxT, TPB_DEFAULT>
-      <<<grid, TPB_DEFAULT, smemSize, s>>>(hist,
-                                           params.n_bins,
-                                           params.max_depth,
-                                           params.min_samples_split,
-                                           params.max_leaves,
-                                           input,
-                                           curr_nodes,
-                                           col,
-                                           done_count,
-                                           mutex,
-                                           splits,
-                                           objective,
-                                           treeid,
-                                           workload_info,
-                                           seed);
-    ML::POP_RANGE();  // computeSplitKernel
+      <<<grid, TPB_DEFAULT, smemSize, handle.get_stream()>>>(hist,
+                                                             params.n_bins,
+                                                             params.max_depth,
+                                                             params.min_samples_split,
+                                                             params.max_leaves,
+                                                             input,
+                                                             d_work_items,
+                                                             col,
+                                                             done_count,
+                                                             mutex,
+                                                             splits,
+                                                             objective,
+                                                             treeid,
+                                                             workload_info,
+                                                             seed);
+    ML::POP_RANGE();  // computeSplitClassificationKernel
     ML::POP_RANGE();  // Builder::computeSplit
   }
+
+  // Set the leaf value predictions in batch
+  void SetLeafPredictions(std::vector<NodeT>* tree)
+  {
+    MLCommon::device_buffer<NodeT> d_tree(
```

Comment:
As I understand, we are allocating a device buffer of size equal to complete tree. This is on top of the builder workspace. Which is problematic in two ways,
1. We are not allocating all the device space at one go
2. The device allocation is variable of variable size

Both issues could be fixed if we reuse the builder workspace and call the `leafKernel` in batched manner.

Reply:
I disagree here. This allocation happens once per tree, is it really going to have a performance impact? If the allocation is too large, then FIL would crash anyway. I can't see the size of this particular allocation being a problem. The existing code seems simpler and safer to me.

Reply:
I have batched the computation.